spec/index.bs

```diff
@@ -290,15 +290,41 @@ The {{Navigator/gpu}} attribute is used to access a {{GPU}} object from the main
 
 <script type=idl>
 [Exposed=Window]
-interface GPU {
+interface GPU : EventTarget {
     // May reject with DOMException  // TODO: DOMException("OperationError")?
     Promise<GPUAdapter> requestAdapter(optional GPURequestAdapterOptions options = {});
+
+    attribute EventHandler onadapteradded;
 };
 </script>
 
 <dfn interface>GPU</dfn> defines the interface of `navigator.gpu`, the entry point to WebGPU.
 It exposes {{GPU/requestAdapter()}}, for acquiring [=adapters=].
 
+{{GPU}} has the following event types:
+
+<dl dfn-type=event dfn-for=GPU>
+    : <dfn>adapteradded</dfn>
+    ::
+        A new adapter is available and may be visible via a call to {{GPU/requestAdapter()}}.
+
+        Note:
+        Handling this event is optional.
+        Applications may use this event to, for example:
+        <br>
+        (1) If initialization previously failed ({{GPU/requestAdapter()}} rejected,
+            detect when it should re-attempt initialization.
+        <br>
+        (2) If already using WebGPU, switch to a higher-power adapter if one becomes available.
+
+        Note:
+        A single system event (e.g. battery status change, or GPU connect/disconnect)
+        should never trigger multiple competing signals,
+        so it is safe to handle each such signal individually.
+        (For example, a system event may trigger multiple {{GPUDevice/lost}} signals,
+        but not both {{GPU/adapteradded}} and {{GPUDevice/lost}}.)
```

Comment:
Is a reboot of an external GPU a single event or two?

Reply:
Good question. It would have to be two: adapter loss and adapter added. We would at least need to ensure those events arrive in order.

Even if these are spaced far enough apart that they don't collide, there is a problem: when the adapter comes back, if a promise from getAdapter() is still waiting (e.g. because it's waiting for the reset to finish), then there will be two signals: adapteradded and the resolution of the promise.

Reply:
https://github.com/gpuweb/gpuweb/pull/524#discussion_r359302558 may solve this

Reply:
One way to handle this: don't fire `adapteradded` on device resets. If an app did requestAdapters() after a device reset it would see a new `GPUAdapter` object, but it would not be a different physical device so it doesn't actually care.

Reply:
Actually, as one TODO in the doc says, I don't think there's a reason to actually lose the adapter on reset. We can just lose all the devices on the adapter without replacing the adapter itself. So then there is no reason to expect `adapteradded` would fire anyway.

Reply:
Updated.